addons/binding/org.openhab.binding.modbus/src/main/java/org/openhab/binding/modbus/discovery/internal/ModbusDiscoveryService.java

```diff
@@ -0,0 +1,168 @@
+
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.smarthome.config.discovery.AbstractDiscoveryService;
+import org.eclipse.smarthome.config.discovery.DiscoveryResult;
+import org.eclipse.smarthome.config.discovery.DiscoveryService;
+import org.eclipse.smarthome.core.thing.ThingTypeUID;
+import org.openhab.binding.modbus.discovery.ModbusDiscoveryParticipant;
+import org.openhab.binding.modbus.handler.ModbusEndpointThingHandler;
+import org.osgi.service.component.annotations.Component;
+import org.osgi.service.component.annotations.Reference;
+import org.osgi.service.component.annotations.ReferenceCardinality;
+import org.osgi.service.component.annotations.ReferencePolicy;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ *
+ * Discovery service for Modbus bridges.
+ * New bridges (TCP or Serial Modbus endpoint) should be registered with this service.
+ * Also any bundles that perform auto discovery should register a ModbusDiscoveryParticipant.
+ * This ModbusDiscoveryParticipants will be called by the service when
+ * a discovery scan is requested.
+ *
+ * @author Nagy Attila Gabor - initial contribution
+ *
+ */
+@Component(immediate = true, service = DiscoveryService.class, configurationPid = "discovery.modbus")
+public class ModbusDiscoveryService extends AbstractDiscoveryService {
+
+    private final Logger logger = LoggerFactory.getLogger(ModbusDiscoveryService.class);
+
+    // Set of handlers that allow discovery
+    private final Set<ModbusEndpointThingHandler> handlers = new CopyOnWriteArraySet<>();
+
+    // Set of the registered participants
+    private final Set<ModbusDiscoveryParticipant> participants = new CopyOnWriteArraySet<>();
+
+    // Set of the supported thing types. This is a union of all the thing types
+    // supported by the registered discovery services.
+    private final Set<ThingTypeUID> supportedThingTypes = new CopyOnWriteArraySet<>();
+
+    private static final int SEARCH_TIME = 5;
+
+    // Number of running discovery processes
+    private int discoveryProcessCount = 0;
+
+    /**
+     * Constructor for the discovery service.
+     * Set up default parameters
+     */
+    public ModbusDiscoveryService() {
+        // No supported thing types by default
+        // Search time is for the visual reference
+        // Background discovery disabled by default
+        super(null, SEARCH_TIME, false);
+    }
+
+    /**
+     * Begin a discovery scan over each endpoint
+     */
+    @Override
+    protected void startScan() {
+        logger.trace("ModbusDiscoveryService starting scan");
+        for (ModbusEndpointThingHandler handler : handlers) {
+            if (handler.isDiscoveryEnabled()) {
+                logger.debug("Starting scan for handler {}", handler);
+                new ModbusDiscoveryProcess(this, participants, handler).start();
+                discoveryProcessCount++;
+            }
+        }
+        if (discoveryProcessCount == 0) {
```

Comment:
That seems to be a bit fragile with a counter. What if an osgi service unregisters in the meantime?

Reply:
I'm using this to keep track of the outstanding discovery processes, and get to know, when all of them has been finished, so the stopScan() method can be called on the AbstractDiscoveryService.
I didn't want to keep a reference to the instances, because that could cause a memory leak if they forget to report back.
@davidgraeff would you like it better if each of the ModbusDiscoveryProcesses would get a unique id, and we would keep a set of the running ones? That could be cleared when a new discovery process is started.

Reply:
How about a WeakHashMap?

Reply:
I've sorted this out by moving the discovery tasks into a service class connected to the handler.